util/string_helpers.py

```diff
@@ -0,0 +1,86 @@
+# Helper functions and objects regarding strings -- especially stuff
+# that lets us negotiate the distinction between Unicode and
+# bytestrings.
+
+import base64 as stdlib_base64
+import binascii
+import os
+import sys
+
+class UnicodeAwareBase64(object):
+    """Simulate the interface of the base64 module, but make it look as
+    though base64-encoding and -decoding works on Unicode strings.
+
+    Behind the scenes, Unicode strings are encoded to a particular
+    encoding, then base64-encoded or -decoded, then decoded from that
+    encoding.
+
+    Since we get Unicode strings out of the database, this lets us
+    base64-encode and -decode strings based on those strings, without
+    worrying about encoding to bytes and then decoding.
+    """
+
+    def __init__(self, encoding):
+        self.encoding = encoding
+
+    def _ensure_bytes(self, s):
+        if isinstance(s, bytes):
+            return s
+        return s.encode(self.encoding)
+
+    def _ensure_unicode(self, s):
+        if isinstance(s, bytes):
+            return s.decode(self.encoding)
+        return s
+
+    def wrap(func):
+        def wrapped(self, s, *args, **kwargs):
+            s = self._ensure_bytes(s)
+            value = func(s, *args, **kwargs)
+            return self._ensure_unicode(value)
+        return wrapped
+
+    # Wrap most of the base64 module API so that Unicode is handled
+    # transparently.
+    b64encode = wrap(stdlib_base64.b64encode)
+    b64decode = wrap(stdlib_base64.b64decode)
+    standard_b64encode = wrap(stdlib_base64.standard_b64encode)
+    standard_b64decode = wrap(stdlib_base64.standard_b64decode)
+    urlsafe_b64encode = wrap(stdlib_base64.urlsafe_b64encode)
+    urlsafe_b64decode = wrap(stdlib_base64.urlsafe_b64decode)
+
+    # These are deprecated in base64 and we should stop using them.
+    encodestring = wrap(stdlib_base64.encodestring)
+    decodestring = wrap(stdlib_base64.decodestring)
+    
+# If you're okay with a Unicode strings being converted to/from UTF-8
+# when you try to encode/decode them, you can use this object instead of
+# the standard 'base64' module.
+base64 = UnicodeAwareBase64("utf8")
+
+def random_string(size):
+    """Generate a random string of binary, encoded as hex digits.
+
+    :param: Size of binary string in bytes.
+    :return: A Unicode string.
+    """
+    return binascii.hexlify(os.urandom(size)).decode("utf8")
+
+
+def native_string(x):
```

Comment:
Is this something that could also be used in circulation/core? It'd be nice to share this file between both admins without having to copy/paste the file.

Reply:
This is a copy of a file in core -- we also have copies of log.py and util/http.py, possibly a couple others. There are a couple ways of avoiding this copy-paste. We could have a tiny 'util' submodule that only contained utilities like this, or we could split out the model portion of 'core' into its own submodule (used by circulation and metadata) and have 'core' just be stuff that doesn't depend on any particular model.

The amount of copy-and-pasted code is small enough that I'm not concerned about this, but I do think the model used by circulation and metadata is too monolithic and could be split up.